kclvm/runner/src/assembler.rs

```diff
@@ -0,0 +1,162 @@
+                    }
+                }
+                Err(e) => println!("{:?}", e),
+            };
+        }
+
+        let cache_dir = Path::new(&program.root)
+            .join(".kclvm")
+            .join("cache")
+            .join(kclvm_version::get_full_version());
+        if !cache_dir.exists() {
+            std::fs::create_dir_all(&cache_dir).unwrap();
+        }
+        let mut compile_progs: IndexMap<
+            String,
+            (
+                ast::Program,
+                IndexMap<String, IndexMap<String, String>>,
+                PathBuf,
+            ),
+        > = IndexMap::default();
+        for (pkgpath, modules) in program.pkgs {
+            let mut pkgs = HashMap::new();
+            pkgs.insert(pkgpath.clone(), modules);
+            let compile_prog = ast::Program {
+                root: program.root.clone(),
+                main: program.main.clone(),
+                pkgs,
+                cmd_args: vec![],
+                cmd_overrides: vec![],
+            };
+            compile_progs.insert(
+                pkgpath,
+                (compile_prog, scope.import_names.clone(), cache_dir.clone()),
+            );
+        }
+        let pool = ThreadPool::new(4);
+        let (tx, rx) = channel();
+        let prog_count = compile_progs.len();
+        for (pkgpath, (compile_prog, import_names, cache_dir)) in compile_progs {
+            let tx = tx.clone();
+            pool.execute(move || {
+                let root = &compile_prog.root;
+                let is_main_pkg = pkgpath == kclvm_ast::MAIN_PKG;
+                let file = if is_main_pkg {
+                    PathBuf::from(&pkgpath)
+                } else {
+                    cache_dir.join(&pkgpath)
+                };
+                let ll_file = file.to_str().unwrap();
+                let ll_path = format!("{}.ll", ll_file);
+                let dylib_path = format!("{}{}", ll_file, Command::get_lib_suffix());
+                let mut ll_path_lock =
+                    fslock::LockFile::open(&format!("{}.lock", ll_path)).unwrap();
+                ll_path_lock.lock().unwrap();
+                if Path::new(&ll_path).exists() {
+                    std::fs::remove_file(&ll_path).unwrap();
+                }
+                let dylib_path = if is_main_pkg {
+                    emit_code(
+                        &compile_prog,
+                        import_names,
+                        &EmitOptions {
+                            from_path: None,
+                            emit_path: Some(&ll_file),
+                            no_link: true,
+                        },
+                    )
+                    .expect("Compile KCL to LLVM error");
+                    let mut cmd = Command::new(plugin_agent);
+                    cmd.run_clang_single(&ll_path, &dylib_path)
+                } else {
+                    // If AST module has been modified, ignore the dylib cache
```

Comment:
Why ATS has beed modified here?

Reply:
This comment is left from the previous code and has been corrected.